### PR TITLE
refactored tag_t handling

### PR DESCRIPTION
**Pull request by RalphSteinhagen**
* added source, monitor, and sink example.
* added basic concepts `HasProcessOneFunction<>`, `HasProcessBulkFunction<>`, and   `HasRequiredProcessFunction<>` for non-templated `process_one`  and `process_bulk` to better enforce/signal user-block implementation requirements

This PR only tackles the user-level tag_t propagation API and has still a corner case that needs to be fixed ... as discussed with @ivan-cukic.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -31,7 +31,7 @@ include(FetchContent)
 FetchContent_Declare(
         fmt
         GIT_REPOSITORY https://github.com/fmtlib/fmt.git
-        GIT_TAG 8.1.1
+        GIT_TAG 10.0.0
```
Comment (wirew0rm) [inline CMakeLists.txt:34]: Does this affect the plan to eventually migrate to `std::format` and compatibility to the fmt version used in opencmw? (As we've been bitten by fmt version mismatches before)

Comment (RalphSteinhagen) [reply CMakeLists.txt:34]: Shouldn't ... most of the issues I found were related to sloppy/missing/implicit formatted conversion... having changed to these stricter constraints and moving to std::format should make this rather easier...

**Comment by RalphSteinhagen** [comment]
Thanks a lot to @ivan-cukic and @wirew0rm for fixing/detecting the off-by-one error... :beers: 
```
test/qa_tags
TagSource::process_one(...)      publish tag at       0 @index= 0: {   key: value@0 }
TagSource::process_one(...)      publish tag at       1 @index= 1: {   key: value@1 }
TagSource::process_one(...)      publish tag at     100 @index= 100: {   key: value@100 }
TagSource::process_one(...)      publish tag at     150 @index= 150: {   key: value@150 }
TagSource::process_one(...)      publish tag at    1000 @index= 1000: {   key: value@1000 }
TagSource::process_one(...)      publish tag at    1001 @index= 1001: {   key: value@1001 }
TagSource::process_one(...)      publish tag at    1002 @index= 1002: {   key: value@1002 }
TagSource::process_one(...)      publish tag at    1023 @index= 1023: {   key: value@1023 }
TagMonitor1::process_bulk(...)   received tag at      0 @index= 0: {   key: value@0 }
TagMonitor2::process_one(...)    received tag at      0 @index= 0: {   key: value@0 }
TagSink1::process_bulk(...)      received tag at      0 @index= 0: {   key: value@0 }
TagSink2::process_one(...)       received tag at      0 @index= 0: {   key: value@0 }
TagMonitor1::process_bulk(...)   received tag at      1 @index= 0: {   key: value@1 }
TagMonitor2::process_one(...)    received tag at      1 @index= 0: {   key: value@1 }
TagSink1::process_bulk(...)      received tag at      1 @index= 0: {   key: value@1 }
TagSink2::process_one(...)       received tag at      1 @index= 0: {   key: value@1 }
TagMonitor1::process_bulk(...)   received tag at    100 @index= 0: {   key: value@100 }
TagMonitor1::process_bulk(...)   received tag at    150 @index= 0: {   key: value@150 }
TagMonitor2::process_one(...)    received tag at    100 @index= 0: {   key: value@100 }
TagSink1::process_bulk(...)      received tag at    100 @index= 0: {   key: value@100 }
TagSink2::process_one(...)       received tag at    100 @index= 0: {   key: value@100 }
TagMonitor1::process_bulk(...)   received tag at   1000 @index= 0: {   key: value@1000 }
TagMonitor1::process_bulk(...)   received tag at   1001 @index= 0: {   key: value@1001 }
TagMonitor2::process_one(...)    received tag at    150 @index= 0: {   key: value@150 }
TagSink1::process_bulk(...)      received tag at    150 @index= 0: {   key: value@150 }
TagSink2::process_one(...)       received tag at    150 @index= 0: {   key: value@150 }
TagMonitor1::process_bulk(...)   received tag at   1002 @index= 0: {   key: value@1002 }
TagMonitor1::process_bulk(...)   received tag at   1023 @index= 0: {   key: value@1023 }
TagMonitor2::process_one(...)    received tag at   1000 @index= 0: {   key: value@1000 }
TagSink1::process_bulk(...)      received tag at   1000 @index= 0: {   key: value@1000 }
TagSink2::process_one(...)       received tag at   1000 @index= 0: {   key: value@1000 }
TagMonitor2::process_one(...)    received tag at   1001 @index= 0: {   key: value@1001 }
TagSink1::process_bulk(...)      received tag at   1001 @index= 0: {   key: value@1001 }
TagSink2::process_one(...)       received tag at   1001 @index= 0: {   key: value@1001 }
TagMonitor2::process_one(...)    received tag at   1002 @index= 0: {   key: value@1002 }
TagSink1::process_bulk(...)      received tag at   1002 @index= 0: {   key: value@1002 }
TagSink2::process_one(...)       received tag at   1002 @index= 0: {   key: value@1002 }
TagMonitor2::process_one(...)    received tag at   1023 @index= 0: {   key: value@1023 }
TagSink1::process_bulk(...)      received tag at   1023 @index= 0: {   key: value@1023 }
TagSink2::process_one(...)       received tag at   1023 @index= 0: {   key: value@1023 }
All tests passed (18 asserts in 3 tests)
```

Need to refactor work to make these checks easier ... see also:
  * #81

**Review by wirew0rm — APPROVED** [review]
:+1:  looks good, very nice unittesting.

**Pull request merged**